pkg/cable/wireguard/driver.go

```diff
@@ -0,0 +1,411 @@
+
+	// Add routes to peer
+	//TODO save old routes for removal
+	idx := w.link.Attrs().Index
+	for _, peerNet := range allowedIPs {
+		route := netlink.Route{
+			LinkIndex: idx,
+			Dst:       &peerNet,
+		}
+		if err = netlink.RouteAdd(&route); err != nil {
+			return "", fmt.Errorf("failed to add route %s: %v", route, err)
+		}
+	}
+
+	klog.V(log.DEBUG).Infof("Done connecting endpoint peer %+v", peerCfg)
+	return ip, nil
+}
+
+func publicKey(ep types.SubmarinerEndpoint) (*wgtypes.Key, error) {
+	s, found := ep.Spec.BackendConfig[PublicKey]
+	if !found {
+		return nil, fmt.Errorf("endpoint is missing public key")
+	}
+	key, err := wgtypes.ParseKey(s)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse public key %s: %v", s, err)
+	}
+	return &key, nil
+}
+
+func (w *wireguard) DisconnectFromEndpoint(remoteEndpoint types.SubmarinerEndpoint) error {
+	klog.V(log.DEBUG).Infof("Removing endpoint %v+", remoteEndpoint)
+
+	if w.localEndpoint.Spec.ClusterID == remoteEndpoint.Spec.ClusterID {
+		klog.V(log.DEBUG).Infof("Will not disconnect self")
+		return nil
+	}
+
+	// parse remote public key
+	remoteKey, err := publicKey(remoteEndpoint)
+	if err != nil {
+		return fmt.Errorf("failed to parse peer public key: %v", err)
+	}
+
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+
+	// key mismatch indicates removal of old peer key, where clusterID is already associated with new key or no key
+	keyMismatch := false
+
+	oldKey, found := w.peers[remoteEndpoint.Spec.ClusterID]
+	if !found {
+		keyMismatch = true
+		klog.Warningf("Key mismatch, cluster %s has no key but asked to remove %s", remoteEndpoint.Spec.ClusterID, remoteKey)
+	} else if oldKey.String() != remoteKey.String() {
+		keyMismatch = true
+		klog.Warningf("Key mismatch, cluster %s key is %s but asked to remove %s", remoteEndpoint.Spec.ClusterID, oldKey, remoteKey)
+	}
+
+	// wg remove
+	_ = removePeer(w.client, remoteKey)
+
+	if keyMismatch {
+		// ClusterID probably already associated with new key. Do not remove peers entry nor routes
+		klog.Warningf("Key mismatch for peer cluster %s, keeping existing routes and public key", remoteEndpoint.Spec.ClusterID)
+		return nil
+	}
+	// no delete on mismatch
+	delete(w.peers, remoteEndpoint.Spec.ClusterID)
+
+	// del routes
+	allowedIPs := parseSubnets(remoteEndpoint.Spec.Subnets)
+	idx := w.link.Attrs().Index
+	for _, peerNet := range allowedIPs {
+		route := netlink.Route{
+			LinkIndex: idx,
+			Dst:       &peerNet,
+		}
+		if err = netlink.RouteDel(&route); err != nil {
+			return fmt.Errorf("failed to delete route %s: %v", route, err)
+		}
+	}
+
+	klog.V(log.DEBUG).Infof("Done removing endpoint for cluster %s", remoteEndpoint.Spec.ClusterID)
+	return nil
+}
+
+func (w *wireguard) GetActiveConnections(clusterID string) ([]string, error) {
+	// force caller to skip duplicate handling
+	return make([]string, 0), nil
+}
+
+// Create new wg link and assign addr from local subnets
+func setWGLink(w *wireguard, localSubnets []string) error {
```

Comment:
minor thing but you could `w *wireguard` the receiver instead of explicitly passing in.

Reply:
done